[GEOS-10123] - Internationalization for title and abstract (#5122)
geoserver/geoserver: doc/en/user/source/services/internationalization/index.rst

.. _internationalization:

Internationalization (i18n)
===========================

> GeoServer provides the possibility to return the content of a GetCapabilities response according to user defined languages. The functionality is available for the following services:

Author: fixed

* WMS 1.1 and 1.3
* WFS 2.0
* WCS 2.0

Configuration
=============

GeoServer provides an i18n editor for the title and abstract of:

* Layers configuration page.
* Layergroups configuration page.
* WMS, WFS, WCS service configuration pages.
* For Styles i18n configuration see :ref:`sld_language`.

The editor is disabled by default and can be enabled from the i18n checkbox:

.. figure:: img/i18nUIEditor.png

GetCapabilities
===============

> The response content language can be defined through the ``AcceptLanguages`` request param. The way in which GeoServer will reply to the parameter value follows the below rules:

Author: fixed

> * the content being returned according to the specified language will comprise titles, abstracts and keywords.

Author: fixed

> * if a single language code is specified, eg. ``AcceptLanguages=en`` GeoServer will try to return the content in that language. If no content is found in that language an error message will be returned.

Author: fixed

> * if multiple language codes are specified, eg. ``AcceptLanguages=en fr`` for each internationalizable content GeoServer will try to return it in one of the specified language. If no content is found for each language an error message will be returned.

Author: fixed

> * languages can be configured and request also according to local language variants eg. ``AcceptLanguages=en fr-CA``. If any i18n content has been specified with a local variant eg. ``fr-CA`` and the request parameters specifies only the language code eg. ``AcceptLanguages=fr``, the ``fr-CA`` content will be encoded in the response, while in the inverse case the content will not be included.

Author: Did not noticed this when I checked first time...seems that both are supported, I did not find any more precise information about this. I've changed the code so that it can handle both the space and the , separators.

* if a ``*`` is present among the parameter values, eg ``AcceptLanguages=en fr *``, GeoServer will try to return the content in one of the specified language code. If no content is found content will be returned in a language among the ones availables.

> * if not all the configurable elements have i18n title and abstract available for the requested language, GeoServer will encode those attributes only for services, layers, layergroups and styles that have them defined and will put, in place of the missing internazionaled content an error message like the following: ``DID NOT FIND i18n CONTENT FOR THIS ELEMENT`` as in the example below.

Author: indeed my fault, fixed.

.. _internationalization:

Internationalization (i18n)
===========================

GeoServer supports returning a GetCapabilities document in various languages. The functionality is available for the following services:

* WMS 1.1 and 1.3
* WFS 2.0
* WCS 2.0

Configuration
=============

GeoServer provides an i18n editor for the title and abstract of:

* Layers configuration page.
* Layergroups configuration page.
* WMS, WFS, WCS service configuration pages.
* For Styles i18n configuration see :ref:`sld_language`.

The editor is disabled by default and can be enabled from the i18n checkbox:

.. figure:: img/i18nEditor.png

GetCapabilities
===============

> The response content language can be selected using the ``AcceptLanguages`` request parameter. The GeoServer response will vary based on the below rules:

Reviewer: ```suggestion
The response content language can be selected using the ``AcceptLanguages`` request parameter. The GeoServer response will vary based on the following rules:
```

Author: fixed

* If multiple language codes are specified, e.g. ``AcceptLanguages=en fr`` or ``AcceptLanguages=en,fr``, for each internationalizable content GeoServer will try to return it in one of the specified language. If no content is found for each language an error message will be returned.

* Languages can be configured and request also according to local language variants e.g. ``AcceptLanguages=en fr-CA`` or ``AcceptLanguages=en,fr-CA``. If any i18n content has been specified with a local variant eg. ``fr-CA`` and the request parameters specifies only the language code e.g. ``AcceptLanguages=fr``, the ``fr-CA`` content will be encoded in the response, while in the inverse case the content will not be included.

* If a ``*`` is present among the parameter values, e.g. ``AcceptLanguages=en fr *`` or ``AcceptLanguages=en,fr,*``, GeoServer will try to return the content in one of the specified language code. If no content is found content will be returned in a language among the ones availables.

> * If not all the configurable elements have i18n title and abstract available for the requested language, GeoServer will encode those attributes only for services, layers, layergroups and styles that have them defined. In case the missing value is the tile, in place of the missing internazionaled content an error message like the following, will appear: ``DID NOT FIND i18n CONTENT FOR THIS ELEMENT``.

Reviewer: ```suggestion
* If not all the configurable elements have i18n title and abstract available for the requested language, GeoServer will encode those attributes only for services, layers, layergroups and styles that have them defined. In case the missing value is the tile, in place of the missing internationalized content an error message like the following, will appear: ``DID NOT FIND i18n CONTENT FOR THIS ELEMENT``.
```

Author: fixed

.. figure:: img/ErrorMessages.png

Default Language
================

> GeoServer allows to define a default language meant to be used when international content has been set in services', layers' and groups' configuration pages, but no ``AcceptLanguaes`` parameter has been specified in a ``GetCapabilities`` request. The default language can be set from the services' configuration pages (WMS, WFS, WCS) or from global settings from a dropdown like the following:

Reviewer: ```suggestion
GeoServer allows defining a default language to be used when international content has been set in services', layers' and groups' configuration pages, but no ``AcceptLanguages`` parameter has been specified in a ``GetCapabilities`` request. The default language can be set from the services' configuration pages (WMS, WFS, WCS) or from global settings from a dropdown as shown below:
```

Author: fixed

.. figure:: img/DefaultLanguage.png

> GeoServer will first check for a service specific default locale. If not found will use the one in the global settings. If not even the global default language has been specified, GeoServer will use the server default locale.

Reviewer: ```suggestion
GeoServer will first check for a service specific default locale. If not found, it will use the one in the global settings. If  the global default language has not been specified, GeoServer will use the server default locale.
```

Author: fixed
